src/orquestra/sdk/_base/cli/_dorq/_login/_login_server.py

```diff
@@ -0,0 +1,70 @@
+import asyncio
+from typing import Optional
+
+from aiohttp import web
+
+
+class LoginServer:
+    def __init__(
+        self,
+        runtime_repo,
+        presenter,
+    ):
+        self._runtime_repo = runtime_repo
+        self._presenter = presenter
+        self._token: Optional[str] = None
+        self._login_url: Optional[str] = None
+
+    @property
+    def token(self):
+        return self._token
+
+    @property
+    def login_url(self):
+        return self._login_url
+
+    async def run(
+        self,
+        cluster_url: str,
+        is_ce: bool,
+        listen_host: str,
+        timeout: int,
+    ):
+        app = web.Application()
+        app["cluster_url"] = cluster_url
+        app.add_routes([web.get("/state", self._state_handler)])
+        runner = web.AppRunner(app)
+        await runner.setup()
+        try:
+            site = web.TCPSite(runner, listen_host, 0)
+            await site.start()
+            self._login_url = self._runtime_repo.get_login_url(
+                cluster_url, is_ce, runner.addresses[0][1]
```

Comment:
**praise**: I really like the fact that it doesn't hardcode the port in the callback URL!